lib/improver/spotdata/neighbour_finding.py

```diff
@@ -0,0 +1,307 @@
+        conditions.
+
+        A vertical displacement bias may be applied with the vertical_bias
+        keyword; whether to prefer grid points above or below the site, or
+        neither.
+
+        A land constraint may be applied that requires a land grid point be
+        selected for a site that is over land. Currently this is established
+        by checking that the nearest grid point barring any other conditions
+        is a land point. If a site is a sea point it will use the nearest
+        neighbour as there should be no vertical displacement difference with
+        other sea points.
+
+        Args:
+        -----
+        cube/sites : See process() above.
+
+        default_neighbours : numpy.array
+            An existing list of neighbours from which variations are made using
+            specified options (e.g. land_constraint). If unset the
+            fast_nearest_neighbour method will be used to build this list.
+
+        orography : numpy.array
+            Array of orography data extracted from an iris.cube.Cube that
+            corresponds to the grids on which all other input diagnostics
+            will be provided.
+
+        land_mask : numpy.array
+            Array of land_mask data extracted from an iris.cube.Cube that
+            corresponds to the grids on which all other input diagnostics
+            will be provided.
+
+        no_neighbours : int
+            Number of grid points about the site to consider when relaxing the
+            nearest neighbour condition. If unset this defaults to 9.
+            e.g. consider a 5x5 grid of points -> no_neighbours = 25.
+
+        Returns:
+        --------
+        neighbours: See process() above.
+
+        """
+
+        # Use the default nearest neighbour list as a starting point, and
+        # if for some reason it is missing, recreate the list using the fast
+        # method.
+        if default_neighbours is None:
+            neighbours = self.fast_nearest_neighbour(cube, sites,
+                                                     orography=orography)
+        else:
+            neighbours = default_neighbours
+
+        for i_site, site in enumerate(sites.itervalues()):
+
+            altitude = site['altitude']
+
+            i, j, dz_nearest = (neighbours['i'][i_site],
+                                neighbours['j'][i_site],
+                                neighbours['dz'][i_site])
+            edgepoint = neighbours['edgepoint'][i_site]
+
+            node_list = nearest_n_neighbours(i, j, no_neighbours)
+            if edgepoint:
+                node_list = node_edge_test(node_list, cube)
+
+            if self.land_constraint:
+                # Check that we are considering a land point and that at least
+                # one neighbouring point is also land. If not no modification
+                # is made to the nearest neighbour coordinates.
+
+                exclude_self = nearest_n_neighbours(i, j, no_neighbours,
```

Comment:
It might be better to avoid naming this variable the same as the `exclude_self` optional argument within `nearest_n_neighbours`.